server/service/src/sync/file_sync_driver.rs

```diff
@@ -0,0 +1,196 @@
+use std::sync::Arc;
+
+use crate::{
+    service_provider::ServiceProvider, settings::Settings, static_files::StaticFileService,
+    sync::file_synchroniser::FileSynchroniser,
+};
+
+use super::settings::SyncSettings;
+use tokio::{
+    sync::mpsc::{self, Receiver, Sender},
+    time::Duration,
+};
+use util::is_central_server;
+
+const FILE_SYNC_UPLOAD_DELAY: Duration = Duration::from_millis(100); // This just gives time for a PAUSE message to be received between uploading files
+const FILE_SYNC_NO_FILES_DELAY: Duration = Duration::from_millis(10000); // If there's nothing to upload or there was an error, wait a longer before checking again
+
+pub enum FileSyncMessage {
+    Start,   // Start sync (could be manual trigger, or automatic on server startup)
+    Stop,    // Stop sync (could be manual trigger, or automatic on server shutdown)
+    Pause, // Pause sync this is called by the main sync process to pause the file sync during a normal sync operation
+    UnPause, // Restart sync if it's not Stopped
+}
+
+pub struct FileSyncDriver {
+    receiver: Receiver<FileSyncMessage>,
+    static_file_service: Arc<StaticFileService>,
+}
+
+#[derive(Clone)]
+pub struct FileSyncTrigger {
+    sender: Sender<FileSyncMessage>,
+}
+
+/// Used to 'drive' file sync synchronisation, it's tasks:
+/// * Expose channel for manually triggering sync
+/// * Trigger sync every SyncSettings.interval_seconds (only when initialised)
+impl FileSyncDriver {
+    pub fn init(settings: &Settings) -> (FileSyncTrigger, FileSyncDriver) {
+        // We use a single-element channel so that we can only have one trigger message processed at a time (no need to queue them up?)
+        let (sender, receiver) = mpsc::channel(1);
+
+        let static_file_service = Arc::new(
+            StaticFileService::new(&settings.server.base_dir)
+                .expect("Failed to create static file service"),
+        );
+
+        (
+            FileSyncTrigger { sender },
+            FileSyncDriver {
+                receiver,
+                static_file_service,
+            },
+        )
+    }
+
+    /// FileSyncDriver entry point, this method is meant to be run within main `select!` macro
+    /// should fail only when database is not accessible or when all receivers were dropped
+    ///
+    ///
+    /// Operations:
+    /// * loop
+    ///    * If initialised await for  trigger OR interval sec timeout
+    ///    * If not initialised await only for start trigger
+    ///    * do sync if any of the above were triggered
+    pub async fn run(mut self, service_provider: Arc<ServiceProvider>) {
+        // Default to a paused so file sync should un-pause once the first `sync` completes
+        let mut stopped = false;
```

Comment:
I've added this `stopped` state as well as `pause` as conceptually we might want to have a stop file sync button that users can use if the want to disable uploads due to network connection issues or something?

Reply:
Yes, exactly if the user stops it we don't want to start it automatically when the sync finishes